design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+
+The `ContainerName` field in the `VirtualChunkContainer` is how the users refer to them to pass credentials.
+
+Several types of credentials are possible.
+
+```rust
+pub enum ObjectStoreCredentials {
+    FromEnv,
+    Anonymous,
+    Static {
+        access_key_id: String,
+        secret_access_key: String,
+        session_token: Option<String>,
+    },
+}
+```
+
+An we will have to add at least one more variant once we implement credentials refresh.
+
+### Use case examples
+
+#### Basic usage
+
+All virtual chunks point to S3 us-east-1, single credential set.
+
+* When calling `set_virtual_ref` locations are passed with `s3://` protocol.
+* No persistent configuration needs to be set in the repo.
+* At runtime, the repo instance can be configured passing credentials for container "s3", the default container.
+* If the repo data is also in s3 us-east-1, and the same credentials can be used, the previous point is not needed, the default will be using the same credentials as the ones for the repo.
+
+#### Non default region
+
+All virtual chunks point to S3 sa-east-1, single credential set.
+
+* When calling `set_virtual_ref` locations are passed with `s3://` protocol.
+* A persistent configuration can be set creating a virtual chunk container with prefix `s3://` and region "sa-east-1", this overrides the default s3 container.
+* Alternatively, this config can also be set at runtime, but every reader needs to remember to do it if they want to access virtual chunks
+* At runtime, the repo instance can be configured passing virtual chunk credentials for the container with prefix "s3://".
+* If Icechunk native data is also in s3 sa-east-1, and the same credentials can be used, the previous point is not needed, the default will be using the same credentials as the ones for the repo.
+
+#### Repo on s3 virtual chunks on gcs
+
+The repo is stored in S3 us-east-1, all virtual chunks point to Google cloud storage africa-south1
+
+* When calling `set_virtual_ref` locations are passed with `gcs://` protocol.
+* A persistent configuration can be set creating a virtual chunk container with name gcs and region "africa-south1"
+* Credentials must be set at runtime for the custom protocol
+
+#### Go to local MinIO for testing
+
+All virtual chunks point to S3 us-east-1, for testing purposes the chunks for an array in `s3://bucket/prefix/array` are also copied locally and served with MinIO on port 8080.
+
+* When calling `set_virtual_ref` locations are passed with `s3://` protocol.
+* The icechunk repo instance is initialized with a configuration that creates a container with `name=minio-override`, `prefix=s3://bucket/prefix/array` and `endpoint_url=localhost:9000`
```

Comment:
```suggestion
* The icechunk repo instance is initialized with a configuration that creates a container with `name=minio-override`, `prefix=s3://bucket/prefix/array` and `endpoint_url=localhost:8080`
```

Reply:
9000 is MinIO default port, I'll fix the place where I used 8080 for the example.